### DISCUSSION TITLE
Problems with replacing `depends_on :x11`

### DISCUSSION THREAD

**Discussion by jgerken** [Homebrew/discussions#679] [Tap maintenance and Homebrew development]
I have a formula that depends on X11 and still uses `depends_on :x11` which is no longer allowed. According to the message, I should replace `:x11` with a dependency on a specific X11 formula. Since I am no developer of the underlying project, I was not sure which formulas are all required so I started with `depends_on libx11` only as a first try. However, the build fails with a lot of standard X11 identifiers being unknown (e.g. `XDrawLine`) which are part of the `libx11` formula (`XDrawLine` is even used in the test for the formula).

Could you please tell me, what I am doing wrong? Should I have changed more in the formula then just the `depends_on`? 

I using macOS Catalina (10.15.7) and Homebrew version is:

```
brew --version
Homebrew 2.7.7-120-gd2430bc
Homebrew/homebrew-core (git revision 9700fd; last commit 2021-02-02)
Homebrew/homebrew-cask (git revision 6008e86; last commit 2021-02-03)
```

**Comment by SMillerDev** [comment]
If the software you're building specifically depends on the macOS version of X11 you might have to tell it where to find these identifiers now. Without knowing anything about the software this is impossible to debug though. You're essentially asking "how do I build something with X11"

**Comment by jgerken** [comment]
The formula worked fine as long as XQuartz was installed and it depended on ˋ:x11ˋ. Now the latter is no longer allowed by Homebrew and I thought that depending on ˋlibx11ˋ instead would give a similar result and make parts of XQuartz available to the build. If I need to manually add the XQuartz stuff to the includes in my formula what is the purpose of the new ˋlibx11ˋ formula then? From the error message Homebrew gave me, I thought I had to replace ˋ:x11ˋ with ˋlibx11ˋ and maybe some others and the build would work as it did. before. Isn’t it the whole purpose of Homebrew to make the dependencies available to the build? So, I do not understand why I now should point my build to something it used to get from Homebrew previously.

**Reply by SMillerDev** [reply]
> The formula worked fine as long as XQuartz was installed and it depended on ˋ:x11ˋ. 

This only defined that it needed the app and did nothing else.

> Now the latter is no longer allowed by Homebrew and I thought that depending on ˋlibx11ˋ instead would give a similar result and make parts of XQuartz available to the build.

It makes X11 libs available to the build, XQuartz is an implementation of an X11 server and packages everything X11 could ever want/need.

> If I need to manually add the XQuartz stuff to the includes in my formula what is the purpose of the new ˋlibx11ˋ formula then?

To allow formula to depend on the parts of X11 that they need so that this implementation could be shared between macOS and linux instead of just depending on some installed macOS app.

> From the error message Homebrew gave me, I thought I had to replace ˋ:x11ˋ with ˋlibx11ˋ and maybe some others and the build would work as it did before. 

For almost all software this has been the case, but it seems your software is an exception that has a bunch of hardcoded macOS requirements.

> Isn’t it the whole purpose of Homebrew to make the dependencies available to the build? So, I do not understand why I now should point my build to something it used to get from Homebrew previously.

It is, but homebrew never provided information before. It simply made sure that an app was installed.

**Reply by tlk** [reply]
I have also been battling with a formula that depends on X11.

Apparently, it was decided at some point by the Homebrew project that X11 on macOS is bad. The most comprehensive [justification](https://docs.brew.sh/Acceptable-Formulae#stuff-that-builds-a-gui-by-default-but-doesnt-have-to) I have found is this:

> Don’t build X11/XQuartz GUIs as they are a bad user experience on macOS.

Fair enough, I guess.

You may find it helpful to know that Homebrew actively filters out certain X11 paths during the build process. See https://github.com/Homebrew/brew/blob/master/Library/Homebrew/shims/super/cc#L252-L254

This may confuse libtool/autotools based projects during the `configure` step.

I found that while adding depends_on libxext and libxt would allow the project (Imagemagick) to build it would segfault at runtime. A rather hackish (but functional) alternative to adding new dependencies was to simply add a symlink to X11 in the buildpath to work around the filtering:

```diff
+    ln_s "/opt/X11/include/X11", "#{buildpath}/X11"
```

See https://github.com/tlk/homebrew-imagemagick-x11/blob/784b3d73c7b0141bcd9b7591fb0f41a46ef1e4c6/Formula/imagemagick.rb#L58-L82

**Reply by tlk** [reply]
> > The formula worked fine as long as XQuartz was installed and it depended on ˋ:x11ˋ.
> 
> This only defined that it needed the app and did nothing else.

@SMillerDev, I don't think this is correct.

Considering [this change](https://github.com/Homebrew/brew/commit/87dd13aea6691e9d5e0f3ba8d1d1f862a809212a#diff-ed4c0c77ea1a192666ae8a1758389fb65054649d71e0e39e14ee08e919fdb2b3L13-L34), I believe that `depends_on :x11` would also adjust a number of compiler flags.

However, if you set those flags manually in a formula they may be filtered out silently: https://github.com/Homebrew/brew/blob/master/Library/Homebrew/shims/super/cc#L252-L254

**Comment by carlocab** [comment]
This related issue may help, as it links to a large number of examples of `depends_on :x11` being replaced by dependencies on specific formulae: https://github.com/Homebrew/homebrew-core/issues/64166
